test/units/module_utils/facts/test_date_time.py

```diff
@@ -7,29 +7,45 @@
 
 import pytest
 import datetime
+import time
 
-from ansible.module_utils.facts.system.date_time import DateTimeFactCollector
+from ansible.module_utils.facts.system import date_time
 
-TIMESTAMP = datetime.datetime(2020, 7, 11, 12, 34, 56, 124356)
+EPOCH_TS = 1594449296.123456
+DT = datetime.datetime(2020, 7, 11, 12, 34, 56, 124356)
+DT_UTC = datetime.datetime(2020, 7, 11, 2, 34, 56, 124356)
 
 
 @pytest.fixture
 def fake_now(monkeypatch):
-    """Patch `datetime.datetime.now()` to return a deterministic value."""
+    """
+    Patch `datetime.datetime.fromtimestamp()`, `datetime.datetime.utcfromtimestamp()`,
```

Comment:
I think we don't need to patch `datetime.datetime.fromtimestamp()` and `datetime.datetime.utcfromtimestamp()` methods as they just return a datetime object from patched `time.time()`.

Reply:
+1 it's better to just patch `time.time`

Reply:
I tried not patching those but ended up incorrect `datetime` objects. For example:
```
>       assert date_facts['date_time'][fact_name] == fact_value
E       AssertionError: assert '2020-07-11T06:34:56.123456Z' == '2020-07-11T02:34:56.124356Z'
E         - 2020-07-11T02:34:56.124356Z
E         ?             ^          -
E         + 2020-07-11T06:34:56.123456Z
E         ?             ^         +

```

```
================================================ short test summary info =================================================
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[hour-12]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[epoch-1594434896]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[time-12:34:56]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_basic-20200711T123456124356]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_basic_short-20200711T123456]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_micro-2020-07-11T02:34:56.124356Z]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601-2020-07-11T02:34:56Z]
```

Reply:
@samdoran then, I'd say that there's something wrong with the setup and TZ should be adjusted somewhere.

Reply:
That's what I was thinking. I tried moving the application of the TZ around, but didn't have any luck. I'll keep working on it.

Reply:
So I looked it up and it seems that it hits `time.localtime()` which internals are buried in C:

https://github.com/python/cpython/blob/0275e04/Lib/datetime.py#L829 -> https://github.com/python/cpython/blob/0275e04/Modules/timemodule.c#L497 -> https://github.com/python/cpython/blob/0275e04/Python/pytime.c#L1053 -> https://en.cppreference.com/w/c/chrono/localtime.

It says that `tzset()` will set up internal structures based on `TZ` env var but this seems dangerous and you risk poisoning the entire interpreter state, hence influence other tests if you don't figure out how to clean it up properly.

Looks like it's really safer to monkey-patch methods in the Python-land...